Data usage crawler refactor

klauspost opened the pull request:
## Description

Implementation overview: https://gist.github.com/klauspost/1801c858d5e0df391114436fdad6987b

Includes staticcheck upgrade and fixes for that.

Some quick performance tests of crawls.

This is with a cycle size of 16, 174082 XL objects. ~4K folders at prefix level 2, NVME.

```
BEFORE:
Crawl time 1m17.3025259s
Disk access: 350MB/s
Kernel  Time =    59.468 =   70%
User    Time =    21.984 =   25%
Process Time =    81.453 =   95%    Virtual  Memory =    303 MB

AFTER:
SET MINIO_DISK_USAGE_CRAWL_DELAY=10 (default)
Cycle scan time: 45.354097s
Disk Access: 15MB/s
Kernel  Time =     0.562 =    1%
User    Time =     0.640 =    1%
Process Time =     1.203 =    2%    Virtual  Memory =    304 MB

SET MINIO_DISK_USAGE_CRAWL_DELAY=1
Cycle scan time: 3.3367481s
Disk Access: 160MB/s
Kernel  Time =     0.468 =    3%
User    Time =     1.421 =   12%
Process Time =     1.890 =   16%    Virtual  Memory =    303 MB

SET MINIO_DISK_USAGE_CRAWL_DELAY=0
Cycle scan time: 2.1245395s
Disk Access: 175MB/s
Kernel  Time =     1.500 =   14%
User    Time =     0.765 =    7%
Process Time =     2.265 =   22%    Virtual  Memory =    304 MB
```

## How to test this PR?

For now the server will display extra information on crawling.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] Documentation needed
- [ ] Unit tests needed

vadmeste commented on cmd/storage-rest-server.go:
```diff
@@ -137,17 +138,29 @@ func (s *storageRESTServer) CrawlAndGetDataUsageHandler(w http.ResponseWriter, r
-	usageInfo, err := s.storage.CrawlAndGetDataUsage(GlobalServiceDoneCh)
-	<-doneCh
-
+	w.Header().Set(xhttp.ContentType, "application/binary")
```
I think we are sending `text/event-stream` here so the load balancer will immediately send data as available to clients (although we probably do not expect load balancers between nodes in the same cluster)

klauspost commented on cmd/fs-v1-multipart_test.go:
```diff
@@ -34,33 +35,38 @@ func TestFSCleanupMultipartUploadsInRoutine(t *testing.T) {
 			t.Fatal("Unexpected err: ", err)
 		}
+	} else {
+		t.Error("Item was not cleaned up.")
```
This was missing on previous tests to actually test functionality.

klauspost commented on cmd/fs-v1_test.go:
```diff
@@ -356,8 +356,6 @@ func TestFSListBuckets(t *testing.T) {
 		t.Fatal("Unexpected error: ", err)
 	}
 
-	GlobalServiceDoneCh <- struct{}{}
```
Didn't do anything and just confuse other parts that use it.

klauspost commented on cmd/xl-sets.go:
```diff
@@ -1764,6 +1764,10 @@ func (s *xlSets) healMRFRoutine() {
 	// Wait until background heal state is initialized
 	var bgSeq *healSequence
 	for {
+		if globalBackgroundHealState == nil {
```
This seems to be some global state that isn't always initialized on some tests.

harshavardhana commented on browser/ui-assets.go:
```diff
@@ -21,6 +21,9 @@ import (
 	"time"
 )
 
+//lint:file-ignore ST1003 Generated file...
```
This is an autogenerated file @klauspost we cannot have this in place it will get overwritten next time ui-assets.go are autogenerated

klauspost replied:
See https://github.com/dominikh/go-tools/issues/429 :( 

It seems like we might be able to exclude it only from the 'browser' directory. Let me try that.

klauspost replied:
Added separate config file instead.

harshavardhana commented on cmd/data-usage-cache.go:
```diff
@@ -0,0 +1,357 @@
+package cmd
```
Missing copyright

harshavardhana commented on cmd/storage-rest-server.go:
```diff
@@ -137,17 +138,29 @@ func (s *storageRESTServer) CrawlAndGetDataUsageHandler(w http.ResponseWriter, r
-	usageInfo, err := s.storage.CrawlAndGetDataUsage(GlobalServiceDoneCh)
-	<-doneCh
-
+	w.Header().Set(xhttp.ContentType, "application/binary")
```
We cannot change the contentType here because of requirements to let proxies know that they shouldn't buffer this request since its part of https://www.w3schools.com/html/html5_serversentevents.asp (server sent events) - it was on purpose we kept it `"text/event-stream"`

harshavardhana replied:
If for some reason there are proxies between MinIO servers

klauspost replied:
Reverted.

harshavardhana commented on pkg/s3select/sql/utils.go:
```diff
@@ -50,7 +50,7 @@ func (e *JSONPathElement) String() string {
 }
 
-// Removes double quotes in quoted identifiers
+// String removes double quotes in quoted identifiers
```
Can we not fix the staticcheck updates in another PR instead of this change? - since its a rewrite with a large change, it looks quite intrusive as its touching unexpected parts.

klauspost replied:
I need to update to fix staticcheck failures with generated code.

harshavardhana commented on cmd/fastwalk.go:
```diff
@@ -11,173 +11,12 @@ package cmd
 import (
```
now that we don't do what fastwalk does, does it make sense to keep this code? @klauspost - perhaps move this to the new data-usage file? or rename this file as some `data-usage-utils.go` ?

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -195,45 +200,225 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
+		disks = append(disks, d)
+	}
+	if len(disks) == 0 || len(buckets) == 0 {
+		return nil
```
if all disks are offline shouldn't we log something that as concerning?

klauspost replied:
Not sure if this is something the crawler should care about. I assume there is already plenty of other code complaining if this is true :)

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -195,45 +200,225 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
+			logger.LogIf(ctx, err)
+			oldCache = dataUsageCache{}
+		}
+		//fmt.Println("old cache deserialized:", oldCache.StringAll())
```
perhaps you wanted to add `if dataUsageDebug` ? - not sure if we should keep these `fmt.Println` in the code if there is no way to enable them?

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -195,45 +200,225 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
+		if e != nil {
+			bucketCh <- b
+			cache.replace(b.Name, "/", *e)
+			// fmt.Println("Existing bucket", b.Name)
```
Same question as above

harshavardhana commented on cmd/xl-v1.go:
```diff
@@ -195,45 +200,225 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
+		e := oldCache.find(b.Name)
+		if e != nil {
+			bucketCh <- b
+			cache.replace(b.Name, "/", *e)
```
All "/" can be SlashSeparator constant defined globally.

klauspost replied:
Sounds good - replacing them.

klauspost replied:
Made it a separate constant for clarity.

harshavardhana commented on cmd/xl-zones.go:
```diff
@@ -215,44 +216,94 @@ func (z *xlZones) StorageInfo(ctx context.Context, local bool) StorageInfo {
+
 	wg.Wait()
+	if dataUsageDebug {
+		fmt.Println("Scan time:", time.Since(t))
```
Rather than `fmt.Println` we could use `logger.Info`

klauspost replied:
Cool. Will do 👍

klauspost replied:
Should I drop the `dataUsageDebug`. I suppose this is filtered under normal operation?

harshavardhana replied:
Yes we can drop for now..

vadmeste commented on cmd/xl-zones.go:
```diff
@@ -215,44 +217,94 @@ func (z *xlZones) StorageInfo(ctx context.Context, local bool) StorageInfo {
+
+			// We need to merge since we will get the same buckets from each zone.
+			// Therefore to get the exact bucket sizes we must merge before we can convert.
+			allMerged := dataUsageCache{Info: dataUsageCacheInfo{Name: "/"}}
```
Shouldn't this be a bucket name instead of '/' ?

klauspost replied:
Each zone returns total for each bucket in the zone, so the root of the cache is `/` (changing it to `dataUsageRoot`).

We need to merge buckets individually so we can get the total size across all zones for the size histogram.

vadmeste commented on cmd/xl-v1.go:
```diff
@@ -195,45 +201,222 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
+}
+
+// CrawlAndGetDataUsage will start crawling buckets and send updated totals as they are traversed.
+// Updates are sent on a regular basis and the caller *must* consume them.
```
*must* consume them but what would happen otherwise ?

klauspost replied:
The called function locks up? With a channel I would suppose that is clear?

vadmeste commented on cmd/xl-v1.go:
```diff
@@ -195,45 +201,222 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
+	buckets = nil
+	bucketResults := make(chan dataUsageEntryInfo, len(disks))
 
+	save := func(cache *dataUsageCache, name string) error {
```
This can be moved outside with a clear name like, saveDataUsageCacheInBackend()

klauspost replied:
Yeah, it could be a method on `dataUsageCache`. 👍

vadmeste commented on cmd/xl-v1.go:
```diff
@@ -195,45 +201,222 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
+	}
+
+	// New cache..
+	cache := dataUsageCache{
```
A better name of cache would be xlObjDataUsageStore ?

klauspost replied:
For the struct type?

It is not tied to a specific backend, it works on fs as well.

vadmeste commented on cmd/xl-v1.go:
```diff
@@ -195,45 +201,222 @@ func (xl xlObjects) GetMetrics(ctx context.Context) (*Metrics, error) {
+		updates <- cache
+	}()
+
+	// Start one crawler per disk
```
Are we sure we want to crawl all disks here ?

klauspost replied:
Yes. We aim to spread the load as much as possible.

vadmeste reviewed (commented):
Some comments

fwessels commented on cmd/data-usage-cache.go:
```diff
@@ -0,0 +1,419 @@
+	}
+}
+
+// replaceHashed an entry to the cache based on its hash.
```
replaceHashed an entry --> replaceHashed adds an entry

fwessels commented on cmd/data-usage.go:
```diff
@@ -160,52 +157,289 @@ type Item struct {
+	var addDir func(entName string, typ os.FileMode) error
+	var dirStack = []string{f.root, folder}
+
+	addDir = func(entName string, typ os.FileMode) error {
```
Why not simply use := here? Then line 286 is unnecessary ...

klauspost replied:
We are calling it recursively inside the function `err := readDirFn(path.Join(dirStack...), addDir)`.

If we use `:=` it is not defined yet.

fwessels commented on cmd/data-usage.go:
```diff
@@ -160,52 +157,289 @@ type Item struct {
+	if dataUsageDebug {
+		logger.Info(logPrefix+"New folders: %v"+logSuffix, s.newFolders)
+	}
+	// Add add new folders first
```
One Add to much ...

fwessels reviewed (approved):
Nice work, just minor comments, LGTM.

vadmeste commented on cmd/data-usage-cache.go:
```diff
@@ -0,0 +1,419 @@
+		ObjectsTotalSize:      uint64(flat.Size),
+		ObjectsSizesHistogram: flat.ObjSizes.asMap(),
+		BucketsCount:          uint64(len(e.Children)),
+		BucketsSizes:          d.childSizes(*e).asMap(),
```
Apparently BucketsSizes contains an histogram of objects sizes, but what we really need here is a simple mapping between a bucket name and its total size

{"testbucket": 10334432}

klauspost replied:
Well spotted 👍

klauspost replied:
Fixed.

yeungc committed:
Some fix of chinese docs (#9140)

## Description
Clarify disk (硬盘) and node (节点).
Remove the limit (限制) paragraph since there are no max 16 disks limit now.

## Motivation and Context


## How to test this PR?


## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed
- [ ] Functional tests needed (If yes, add [mint](https://github.com/minio/mint) PR # here: )

klauspost committed:
Data usage crawler refactor

Squashed version of/replaces #9075

## Description

Implementation overview: https://gist.github.com/klauspost/1801c858d5e0df391114436fdad6987b

Includes staticcheck upgrade and fixes for that.

Some quick performance tests of crawls.

This is with a cycle size of 16, 174082 XL objects. ~4K folders at prefix level 2, NVME.

```
BEFORE:
Crawl time 1m17.3025259s
Disk access: 350MB/s
Kernel  Time =    59.468 =   70%
User    Time =    21.984 =   25%
Process Time =    81.453 =   95%    Virtual  Memory =    303 MB

AFTER:
SET MINIO_DISK_USAGE_CRAWL_DELAY=10 (default)
Cycle scan time: 45.354097s
Disk Access: 15MB/s
Kernel  Time =     0.562 =    1%
User    Time =     0.640 =    1%
Process Time =     1.203 =    2%    Virtual  Memory =    304 MB

SET MINIO_DISK_USAGE_CRAWL_DELAY=1
Cycle scan time: 3.3367481s
Disk Access: 160MB/s
Kernel  Time =     0.468 =    3%
User    Time =     1.421 =   12%
Process Time =     1.890 =   16%    Virtual  Memory =    303 MB

SET MINIO_DISK_USAGE_CRAWL_DELAY=0
Cycle scan time: 2.1245395s
Disk Access: 175MB/s
Kernel  Time =     1.500 =   14%
User    Time =     0.765 =    7%
Process Time =     2.265 =   22%    Virtual  Memory =    304 MB
```

## How to test this PR?

For now the server will display extra information on crawling.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

harshavardhana committed:
Merge branch 'master' into data-usage-crawler-recfactor

klauspost committed:
Change cache location + Store children hashes as binary array.

klauspost committed:
Merge branch 'data-usage-crawler-recfactor' of github.com:klauspost/minio into data-usage-crawler-recfactor

klauspost committed:
Use correct metabucket prefix.

klauspost committed:
Update documentation.

klauspost committed:
Change start/cycle today to 5 minutes.
